src/anndata/_io/specs/methods.py

```diff
@@ -1028,30 +1034,82 @@ def write_nullable_integer(
     _writer.write_elem(g, "values", v._data, dataset_kwargs=dataset_kwargs)
 
 
+@_REGISTRY.register_write(
+    H5Group, pd.arrays.StringArray, IOSpec("nullable-string-array", "0.1.0")
+)
+@_REGISTRY.register_write(
+    ZarrGroup, pd.arrays.StringArray, IOSpec("nullable-string-array", "0.1.0")
+)
+def write_nullable_string(
+    f: GroupStorageType,
+    k: str,
+    v: pd.arrays.StringArray,
+    *,
+    _writer: Writer,
+    dataset_kwargs: Mapping[str, Any] = MappingProxyType({}),
+):
+    from ..._settings import settings
+
+    if not settings.allow_write_nullable_strings:
+        msg = (
+            "`anndata.settings.allow_write_nullable_strings` is False, "
+            "because writing of `pd.arrays.StringArray` is new "
+            "and not supported in anndata < 0.11, still use by many people. "
+            "Opt-in to writing these arrays by toggling the setting to True."
+        )
+        raise RuntimeError(msg)
+    g = f.require_group(k)
+    if (mask := v.isna()).any():
+        _writer.write_elem(g, "mask", mask, dataset_kwargs=dataset_kwargs)
```

Comment:
If there are no `NA` values, maybe we should give an `ImplicitModificationWarning` here and write out the whole thing as a normal string array?  Not sure.  For `zarr` at least, the semantics around missing arrays for object stores are not amazing.  I have seen 403s, 404s and now this:

```python
In [3]: zarr.open("https://storage.googleapis.com/vitessce-demo-data/maynard-2021/151673.sdata.zarr/images/visium_151673_annotations")
Out[3]: <zarr.hierarchy.Group '/'>

In [4]: zarr.open("https://storage.googleapis.com/vitessce-demo-data/maynard-2021/151673.sdata.zarr/images/visium_151673_annotations/0")
Out[4]: <zarr.core.Array (7, 600, 600) float64>

In [5]: zarr.open("https://storage.googleapis.com/vitessce-demo-data/maynard-2021/151673.sdata.zarr/images/visium_151673_annotations/foo")
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[5], line 1
----> 1 zarr.open("https://storage.googleapis.com/vitessce-demo-data/maynard-2021/151673.sdata.zarr/images/visium_151673_annotations/foo")

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/zarr/convenience.py:129, in open(store, mode, zarr_version, path, **kwargs)
    127         return open_array(_store, mode=mode, **kwargs)
    128     else:
--> 129         return open_group(_store, mode=mode, **kwargs)
    131 else:
    132     if contains_array(_store, path):

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/zarr/hierarchy.py:1587, in open_group(store, mode, cache_attrs, synchronizer, path, chunk_store, storage_options, zarr_version, meta_array)
   1585         if contains_array(store, path=path):
   1586             raise ContainsArrayError(path)
-> 1587         init_group(store, path=path, chunk_store=chunk_store)
   1589 elif mode in ["w-", "x"]:
   1590     if contains_array(store, path=path):

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/zarr/storage.py:682, in init_group(store, overwrite, path, chunk_store)
    679     store["zarr.json"] = store._metadata_class.encode_hierarchy_metadata(None)  # type: ignore
    681 # initialise metadata
--> 682 _init_group_metadata(store=store, overwrite=overwrite, path=path, chunk_store=chunk_store)
    684 if store_version == 3:
    685     # TODO: Should initializing a v3 group also create a corresponding
    686     #       empty folder under data/root/? I think probably not until there
    687     #       is actual data written there.
    688     pass

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/zarr/storage.py:743, in _init_group_metadata(store, overwrite, path, chunk_store)
    741 key = _prefix_to_group_key(store, _path_to_prefix(path))
    742 if hasattr(store, "_metadata_class"):
--> 743     store[key] = store._metadata_class.encode_group_metadata(meta)
    744 else:
    745     store[key] = encode_group_metadata(meta)

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/zarr/storage.py:1470, in FSStore.__setitem__(self, key, value)
   1468     if self.fs.isdir(path):
   1469         self.fs.rm(path, recursive=True)
-> 1470     self.map[key] = value
   1471     self.fs.invalidate_cache(self.fs._parent(path))
   1472 except self.exceptions as e:

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/fsspec/mapping.py:175, in FSMap.__setitem__(self, key, value)
    173 key = self._key_to_str(key)
    174 self.fs.mkdirs(self.fs._parent(key), exist_ok=True)
--> 175 self.fs.pipe_file(key, maybe_convert(value))

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/fsspec/asyn.py:118, in sync_wrapper.<locals>.wrapper(*args, **kwargs)
    115 @functools.wraps(func)
    116 def wrapper(*args, **kwargs):
    117     self = obj or args[0]
--> 118     return sync(self.loop, func, *args, **kwargs)

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/fsspec/asyn.py:103, in sync(loop, func, timeout, *args, **kwargs)
    101     raise FSTimeoutError from return_result
    102 elif isinstance(return_result, BaseException):
--> 103     raise return_result
    104 else:
    105     return return_result

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/fsspec/asyn.py:56, in _runner(event, coro, result, timeout)
     54     coro = asyncio.wait_for(coro, timeout=timeout)
     55 try:
---> 56     result[0] = await coro
     57 except Exception as ex:
     58     result[0] = ex

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/fsspec/asyn.py:408, in AsyncFileSystem._pipe_file(self, path, value, **kwargs)
    407 async def _pipe_file(self, path, value, **kwargs):
--> 408     raise NotImplementedError

NotImplementedError:
```

Reply:
I see that this is similar to some business about `mask` being present or not for `nullable-{boolean,integer}` (so maybe we would never run into this because we don't open `mask` directly) but this feels different than other cases of masked values since we are checking for any NAs here.  Could you explain a bit maybe what is going on there and how it is similar/different to this?

Reply:
> I see that this is similar to some business about mask being present or not

it’s exactly identical, I reuse that code. We could just leave out this check and always write the mask, even if it’s all-unmasked